File: frontend/src/services/inventory.service.ts
import { apiClient } from '@/lib/axios'
import type { Product } from './product.service'

export type MovementType = 'IN' | 'OUT' | 'ADJUSTMENT' | 'RETURN'
export interface InventoryMovement {
  id: string; product_id: string; movement_type: MovementType
  quantity: number; quantity_before: number; quantity_after: number
  reason?: string; performed_by?: string; created_at: string
}
export interface MovementsResponse { data: InventoryMovement[]; total: number; page: number; page_size: number; total_pages: number }

export const inventoryService = {
  adjust: (data: { product_id: string; quantity: number; movement_type: MovementType; reason?: string }) =>
    apiClient.post<{ data: InventoryMovement }>('/inventory/adjust', data).then((r) => r.data.data),

  history: (params?: { page?: number; page_size?: number; product_id?: string }) =>
    apiClient.get<MovementsResponse>('/inventory/history', { params }).then((r) => r.data),

  lowStock: () =>
    apiClient.get<{ data: Product[] }>('/inventory/low-stock').then((r) => r.data.data),
}

File: backend/app/services/inventory_service.py
import uuid
from typing import List, Optional, Tuple
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from app.audit.service import AuditService
from app.core.exceptions import BadRequestException, NotFoundException
from app.models.inventory import InventoryMovement, MovementType
from app.models.product import Product
from app.models.user import User
from app.repositories.inventory_repository import InventoryRepository
from app.repositories.product_repository import ProductRepository
from app.schemas.inventory import InventoryAdjustRequest


class InventoryService:
    def __init__(self, db: AsyncSession):
        self.db = db
        self.repo = InventoryRepository(db)
        self.product_repo = ProductRepository(db)
        self.audit = AuditService(db)

    async def adjust(
        self, data: InventoryAdjustRequest, current_user: User
    ) -> InventoryMovement:
        result = await self.db.execute(
            select(Product)
            .where(Product.id == data.product_id, Product.deleted_at.is_(None))
            .with_for_update()
        )
        product = result.scalar_one_or_none()
        if not product:
            raise NotFoundException("Product", data.product_id)

        new_quantity = product.quantity + data.quantity
        if new_quantity < 0:
            raise BadRequestException(
                f"Adjustment would result in negative stock ({new_quantity}) for '{product.name}'"
            )

        quantity_before = product.quantity
        product.quantity = new_quantity
        await self.db.flush()

        movement = InventoryMovement(
            product_id=product.id,
            movement_type=data.movement_type,
            quantity=abs(data.quantity),
            quantity_before=quantity_before,
            quantity_after=new_quantity,
            reason=data.reason,
            performed_by=current_user.id,
        )
        self.db.add(movement)
        await self.db.flush()
        await self.db.refresh(movement)

        await self.audit.log(
            action="INVENTORY_ADJUST",
            resource="inventory",
            resource_id=str(product.id),
            user_id=current_user.id,
            old_values={"quantity": quantity_before},
            new_values={"quantity": new_quantity, "adjustment": data.quantity},
        )

        return movement

    async def get_history(
        self,
        page: int = 1,
        page_size: int = 20,
        product_id: Optional[uuid.UUID] = None,
    ) -> Tuple[List[InventoryMovement], int]:
        return await self.repo.get_history(
            offset=(page - 1) * page_size,
            limit=page_size,
            product_id=product_id,
        )

    async def get_low_stock(self) -> List[Product]:
        return await self.product_repo.get_low_stock()
